xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -190,94 +190,102 @@ public ClientInterceptor buildClientInterceptor(
       config = overrideConfig;
     }
     FaultConfig faultConfig = (FaultConfig) config;
-    Long delayNanos = null;
-    Status abortStatus = null;
-    if (faultConfig.maxActiveFaults() == null
-        || activeFaultCounter.get() < faultConfig.maxActiveFaults()) {
-      Metadata headers = args.getHeaders();
-      if (faultConfig.faultDelay() != null) {
-        delayNanos = determineFaultDelayNanos(faultConfig.faultDelay(), headers);
-      }
-      if (faultConfig.faultAbort() != null) {
-        abortStatus = determineFaultAbortStatus(faultConfig.faultAbort(), headers);
-      }
-    }
-    if (delayNanos == null && abortStatus == null) {
-      return null;
-    }
-    final Long finalDelayNanos = delayNanos;
-    final Status finalAbortStatus = getAbortStatusWithDescription(abortStatus);
 
     final class FaultInjectionInterceptor implements ClientInterceptor {
       @Override
       public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
           final MethodDescriptor<ReqT, RespT> method, final CallOptions callOptions,
           final Channel next) {
-        Executor callExecutor = callOptions.getExecutor();
-        if (callExecutor == null) { // This should never happen in practice because
-          // ManagedChannelImpl.ConfigSelectingClientCall always provides CallOptions with
-          // a callExecutor.
-          // TODO(https://github.com/grpc/grpc-java/issues/7868)
-          callExecutor = MoreExecutors.directExecutor();
+        boolean checkFault = false;
+        if (faultConfig.maxActiveFaults() == null
+            || activeFaultCounter.get() < faultConfig.maxActiveFaults()) {
+          checkFault = faultConfig.faultDelay() != null || faultConfig.faultAbort() != null;
         }
-        if (finalDelayNanos != null) {
-          Supplier<? extends ClientCall<ReqT, RespT>> callSupplier;
-          if (finalAbortStatus != null) {
-            callSupplier = Suppliers.ofInstance(
-                new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor));
-          } else {
-            callSupplier = new Supplier<ClientCall<ReqT, RespT>>() {
-              @Override
-              public ClientCall<ReqT, RespT> get() {
-                return next.newCall(method, callOptions);
-              }
-            };
+        if (!checkFault) {
+          return next.newCall(method, callOptions);
+        }
```

Comment:
The skip logic may be simplified a bit
```java
        // Skip the filter when all controls are effectively disabled.
        if (faultConfig.maxActiveFaults() == null
            || activeFaultCounter.get() < faultConfig.maxActiveFaults()) {
          if (faultConfig.faultDelay() == null && faultConfig.faultAbort() == null) {
            next.newCall(method, callOptions);
          }
        }
```

Reply:
the diff looks weird, I meant is for
![image](https://github.com/user-attachments/assets/b2c61db8-cee4-4c3f-9290-dc44a7421343)

Reply:
If the outermost `if` is false, then we short-circuit as well. It is possible to tweak the logic to avoid the separate bool, but was a bit harder to follow (the diff at least).

Reply:
Ah, missed the fact we circuit-break by default 🤦 
Not too surprising however given the amount of double negatives we're dealing here.
In situation like this I'd usually all this logic to a separate unit-testable method

```java
if (shouldSkipFaultCheck(faultConfig, activeFaultCounter::get)) {
  return next.newCall(method, callOptions);
}
```